stubs/redis/redis/client.pyi

```diff
@@ -355,7 +359,26 @@ class Redis(Generic[_StrType]):
     def set_response_callback(self, command, callback): ...
     def pipeline(self, transaction: bool = ..., shard_hint: Any = ...) -> Pipeline: ...
     def transaction(self, func, *watches, **kwargs): ...
-    def lock(self, name, timeout=..., sleep=..., blocking_timeout=..., lock_class=..., thread_local=...): ...
+    @overload
+    def lock(
+        self,
+        name: _Key,
+        timeout: Optional[float] = ...,
+        sleep: float = ...,
+        blocking_timeout: Optional[float] = ...,
+        lock_class: Optional[Type[Lock]] = ...,
+        thread_local: bool = ...,
+    ) -> Lock: ...
+    @overload
+    def lock(
+        self,
+        name: _Key,
+        timeout: Optional[float] = ...,
+        sleep: float = ...,
+        blocking_timeout: Optional[float] = ...,
+        lock_class: Type[_LockType] = ...,
+        thread_local: bool = ...,
+    ) -> _LockType: ...
```

Comment:
I am sorry for the late reply, but here is what I would suggest:

```suggestion
    @overload
    def lock(
        self,
        name: _Key,
        timeout: Optional[float] = ...,
        sleep: float = ...,
        blocking_timeout: Optional[float] = ...,
        lock_class: None = ...,
        thread_local: bool = ...,
    ) -> Lock: ...
    @overload
    def lock(
        self,
        name: _Key,
        timeout: Optional[float],
        sleep: float,
        blocking_timeout: Optional[float],
        lock_class: Type[_LockType],
        thread_local: bool = ...,
    ) -> _LockType: ...
    @overload
    def lock(
        self,
        name: _Key,
        timeout: Optional[float] = ...,
        sleep: float = ...,
        blocking_timeout: Optional[float] = ...,
        *,
        lock_class: Type[_LockType],
        thread_local: bool = ...,
    ) -> _LockType: ...
```

The first overload covers the case where either `lock_class` is not provided or is provided and is `None`. The other two overloads cover the case where `lock_class` is anything else. Ideally we would want to write the following:

```python
    @overload
    def lock(
        self,
        name: _Key,
        timeout: Optional[float] = ...,
        sleep: float = ...,
        blocking_timeout: Optional[float] = ...,
        lock_class: Type[_LockType],
        thread_local: bool = ...,
    ) -> _LockType: ...
```

Unfortunately this is invalid syntax, since an argument without a default cannot follow an argument with default. This is why we split this overload into two: One where `lock_class` is provided as positional argument (overload 2), and one where it is provided as keyword argument (overload 3).

Reply:
Thanks for the explanation!